src/v/datalake/parquet_writer.cc

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#include "datalake/parquet_writer.h"
+
+#include <arrow/array/array_base.h>
+#include <arrow/chunked_array.h>
+#include <arrow/io/file.h>
+#include <arrow/io/memory.h>
+#include <arrow/status.h>
+#include <arrow/table.h>
+#include <arrow/type.h>
+#include <parquet/arrow/writer.h>
+#include <parquet/properties.h>
+
+#include <iostream>
+#include <memory>
+#include <stdexcept>
+
+namespace datalake {
+
+namespace {
+class iobuf_output_stream : public arrow::io::OutputStream {
+public:
+    iobuf_output_stream() = default;
+
+    ~iobuf_output_stream() override = default;
+
+    // Close the stream cleanly.
+    arrow::Status Close() override;
+
+    // Return the position in this stream
+    arrow::Result<int64_t> Tell() const override;
+
+    // Return whether the stream is closed
+    bool closed() const override;
+
+    arrow::Status Write(const void* data, int64_t nbytes) override;
+
+    // TODO: implement this to avoid copying data multiple times
+    // virtual Status Write(const std::shared_ptr<Buffer>& data);
+
+    // Take the data from the iobuf and clear the internal state.
+    iobuf take_iobuf();
```

Comment:
looks like this should be r-value qualified `iobuf take_iobuf() &&` since `_current_iobuf` will be left in a moved-from state after this call.